x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -338,84 +358,172 @@ public void removeFromCache(final CacheKey cacheKey) {
+
+            } else {
+                completableFuture = new CompletableFuture<>();
+                runningShardsEvictions.put(shardEviction, completableFuture);
+                completableFuture.whenComplete((result, e) -> {
+                    synchronized (shardsEvictionsMutex) {
+                        final CompletableFuture<Integer> removedFuture = runningShardsEvictions.remove(shardEviction);
+                        final boolean removed = pendingShardsEvictions.remove(shardEviction);
+                        assert removedFuture == completableFuture;
+                        assert removed;
+                        assert shardsEvictionsInvariant();
+                    }
+                });
+            }
+        }
+        try {
+            final List<CacheFile> cacheFilesToEvict = new ArrayList<>();
+            cache.forEach((cacheKey, cacheFile) -> {
+                if (shardEviction.matches(cacheKey)) {
+                    cacheFilesToEvict.add(cacheFile);
                 }
-                success = true;
-            } finally {
-                assert success : "shard eviction should be successful: " + shardEviction;
-                if (success == false) {
-                    final boolean added = evictedShards.add(shardEviction);
-                    assert added : shardEviction;
+            });
+            for (CacheFile cacheFile : cacheFilesToEvict) {
+                try {
+                    cache.invalidate(cacheFile.getCacheKey(), cacheFile);
+                } catch (RuntimeException e) {
+                    logger.warn(() -> new ParameterizedMessage("failed to evict cache file {}", cacheFile.getCacheKey()), e);
+                    assert false : e;
                 }
             }
+            logger.debug("shard eviction [{}] processed with [{}] cache files invalidated", shardEviction, cacheFilesToEvict.size());
+            completableFuture.complete(cacheFilesToEvict.size());
+        } catch (Exception e) {
+            completableFuture.completeExceptionally(e);
         }
+        return completableFuture;
+    }
+
+    /**
+     * Processes and waits for all pending shard evictions to complete.
+     */
+    private void processAllPendingShardsEvictions() {
+        final List<CompletableFuture<Integer>> pendingFutures;
+        synchronized (shardsEvictionsMutex) {
+            allowShardsEvictions = false;
+            pendingFutures = new ArrayList<>(runningShardsEvictions.values());
+            assert shardsEvictionsInvariant();
+        }
+        if (pendingFutures.isEmpty() == false) {
+            try {
+                final CountDownLatch latch = new CountDownLatch(pendingFutures.size());
+                pendingFutures.forEach(completableFuture -> completableFuture.whenComplete((integer, throwable) -> latch.countDown()));
+                if (latch.await(shardsEvictionsStopTimeout.duration(), shardsEvictionsStopTimeout.timeUnit()) == false) {
```

Comment:
I am not sure we want a timeout here? I have a hard time reason through what it means when timed out - and if we can safely handle that, we should just avoid the wait altogether (not proposing that). I see value though in first waiting for some seconds, then outputting a warning before waiting indefinitely.

Reply:
I think we can also just do without a timeout, in tests we will interrupt if the shutdown takes too long anyway so just no timeout and logging the interrupted exception is good enough IMO

Reply:
Sure. I went with Henning's suggestion of using a read/write lock for this.